src/rgw/rgw_role.cc

```diff
@@ -0,0 +1,259 @@
+#include <errno.h>
+#include <ctime>
+
+#include "common/errno.h"
+#include "common/Formatter.h"
+#include "common/ceph_json.h"
+#include "common/ceph_time.h"
+#include "rgw_rados.h"
+
+#include "include/types.h"
+#include "rgw_string.h"
+
+#include "rgw_common.h"
+#include "rgw_tools.h"
+#include "rgw_role.h"
+
+#define dout_subsys ceph_subsys_rgw
+
+using namespace std;
+
+const string RGWRole::role_name_oid_prefix = "role_names.";
+const string RGWRole::role_oid_prefix = "roles.";
+
+int RGWRole::store_info(bool exclusive)
+{
+  string oid = get_info_oid_prefix() + id;
+
+  bufferlist bl;
+  ::encode(*this, bl);
+  return rgw_put_system_obj(store, store->get_zone_params().roles_pool, oid,
+                bl.c_str(), bl.length(), exclusive, NULL, real_time(), NULL);
+}
+
+int RGWRole::store_name(bool exclusive)
+{
+  RGWNameToId nameToId;
+  nameToId.obj_id = id;
+
+  string oid = get_names_oid_prefix() + name;
+
+  bufferlist bl;
+  ::encode(nameToId, bl);
+  return rgw_put_system_obj(store, store->get_zone_params().roles_pool, oid,
+              bl.c_str(), bl.length(), exclusive, NULL, real_time(), NULL);
+}
+
+int RGWRole::create(bool exclusive)
+{
+  int ret;
+
+  /* check to see the name is not used */
+  ret = read_id(name, id);
+  if (exclusive && ret == 0) {
+    ldout(cct, 0) << "ERROR: name " << name << " already in use for role id "
+                    << id << dendl;
+    return -EEXIST;
+  } else if ( ret < 0 && ret != -ENOENT) {
+    ldout(cct, 0) << "failed reading role id  " << id << ": "
+                  << cpp_strerror(-ret) << dendl;
+    return ret;
+  }
+
+  /* create unique id */
+  uuid_d new_uuid;
+  char uuid_str[37];
+  new_uuid.generate_random();
+  new_uuid.print(uuid_str);
+  id = uuid_str;
+
+  //arn
+  arn = "arn:aws:iam::role" + path + name;
```

Comment:
@adamemerson : I took a look at the IAM Policy Elements link that you shared with me. It looks like ARN plays an important role while deducing permissions for an entity and it will be worthwhile to discuss the format of the ARN with the team for the reasons mentioned above, and come to a conclusion. 
Are you planning to attend the standup today(Friday) to discuss this. Because I usually dont attend stand-ups on Fridays (its too late for me here), but if you plan to discuss this today, I will attend. Please let me know. Thanks.